text/0000-dma.md

```diff
@@ -0,0 +1,337 @@
+desireable, even if it means reaching for some unsafe here and there (possibly merely for efficiency).
+
+## Using DMA in a Device Driver
+
+Twizzler will provide an interface for making a single Twizzler object accessible to a device by way
+of the `DmaObject` type exposed by the twizzler-driver crate. The DmaObject can be created from any
+Twizzler object, and exposes APIs for ensuring coherence and memory safety. Let's take as example a
+device that has a command ring buffer that is used to submit commands and to indicate when a command
+has been completed. A command in the ring buffer can point to another DMA buffer that is used to
+transfer data, and may look like the following:
+
+```{rust}
+struct Command {
+    op: u32,
+    status: u32,
+    buffer: u64,
+}
+```
+
+The `op` field specifies some operation to perform (send packet, etc.), the `status` field specifies
+the result of the command (say, for example, is set to 1 when the command is completed and must be
+cleared to zero for a command to be processed). Finally, the `buffer` field points to the physical
+address of some buffer. Let's also imagine some mechanism for communicating to the device the head
+of the ring so that we might communicate to the device a collection of new commands to process via
+a write to some MMIO register. For the sake of simplicity, let's assume that the buffer is at most 1
+page long.
+
+Setting up some DMA regions may look like:
+
+```{rust}
+let object = create_new_object();
+let dma = DmaObject::new(object);
+
+let command_ring = dma.slice::<Command, some_command_len>(Access::BiDirectional, DmaOptions::default());
+let buffer = dma.slice::<u8, some_buffer_len>(Access::HostToDevice, DmaOptions::default());
+```
+
+At this point, both `command_ring` and `buffer` have types `DmaRegion<[Command; N]>` and `DmaRegion<[u8;N]>` 
+respectively. We can then use `DmaRegion::pin()` to get a list of physical pages associated with the
+region so that we may program the device to operate on this command ring. Then, submitting a command
+would look like:
+
+```{rust}
+buffer.with_mut(0..0x1000, |buf| {
+    fill_out_buffer(buf);
+});
+// Grab a 'pin'of the buffer, which ensures that the associated physical addresses and IOMMU maps will remain static until the dma object is dropped.
+let buffer_pin = buffer.pin();
+// Get the physical address of the first page.
+let buffer_addr = buffer_pin[0].addr();
+// Fill out a new command.
+command_ring.with_mut(0..1, |ring| {
+    ring[0] = Command::new(buffer_addr);
+});
+increment_head();
+```
+
+A pin object can manually release the pages it refers to, but otherwise the lifetime of pinned
+physical memory is the same as the DmaObject itself. By tying pin lifetime to the DMA object and not
+the pin object reduces management complexity of
+avoiding accidentally programming a device with stale physical addresses.
+
+The `DmaRegion::with_mut` function runs a closure while ensuring (some level of) safety. Before the
```

Comment:
Can you specify clearly the "some level" as in what is guaranteed to happen and what is not?